File: arrays-hashing/topKFrequentElements/topKFrequent.py
from typing import List
import heapq

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = {}
        for num in nums:
            if num in count:
                count[num] += 1
            else:
                count[num] = 1

        # 1st approach: sorting -> O(n log n)
        # sorted_dict = dict(sorted(count.items(), key=lambda item: item[1], reverse=True))
        # return [item[0] for item in sorted_dict[:k]]

        # 2nd approach: min Heap -> O(n log k)

        # minHeap = []
        # for num, frequncy in count.items():
        #     heapq.heappush(minHeap, (frequncy, num))

        #     if (len(minHeap) > k):
        #         heapq.heappop(minHeap)

        # return [num for freq, num in minHeap]

        # 3rd approach: Bucket Sort -> O(n)

        buckets = [[] for _ in range(len(nums) + 1)]
        for num, frequency in count.items():
            buckets[frequency].append(num)

        res = []
        index = 0
        for i in range(len(nums), 0, -1):
            if buckets[i] is not None:
                for item in buckets[i]:
                    if index == k:
                        break

                    res[index] = item
                    index += 1


        return res

File: arrays-hashing/topKFrequentElements/topKFrequent.java
import java.util.*;
// import java.util.stream.Collectors;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> count = new HashMap<>();

        // count and store the frequencies of each number
        for (int i = 0; i < nums.length; i++) {
            int val = count.getOrDefault(nums[i], 0);
            count.put(nums[i], val + 1);
        }

        /*
            1st approach: sorting -> O(n log n)

            List<Map.Entry<Integer, Integer>> list = count.entrySet().stream()
                .sorted(Map.Entry.comparingByValue())
                .collect(Collectors.toList());

            int[] res = new int[k];
            for (int i = 0; i < k; i++)
                res[i] = list.remove(list.size() - 1).getKey();

            return res;
        */

        /*
            2nd approach: heap -> O(n log k)

            PriorityQueue<Map.Entry<Integer, Integer>> minHeap = new PriorityQueue<>(Comparator.comparing(Map.Entry::getValue));
            for (Map.Entry<Integer, Integer> entry : count.entrySet()) {
                minHeap.add(entry);
                if (minHeap.size() > k)
                    minHeap.remove();
            }
            int[] res = new int[k];
            for (int i = 0; i < k; i++)
                res[i] = minHeap.remove().getKey();

            return res;
        */

        /* 3rd approach: Bucket Sort -> O(n) */
        List<List<Integer>> buckets = new ArrayList<>(nums.length + 1);

        for (int i = 0; i <= nums.length; i++)
            buckets.add(new ArrayList<>());

        for (Map.Entry<Integer, Integer> entry : count.entrySet()) {
            int frequency = entry.getValue();
            buckets.get(frequency).add(entry.getKey());
        }

        int[] res = new int[k];
        int index = 0;
        for (int i = nums.length; i > 0; i--) {
            if (!buckets.get(i).isEmpty()) {
                for (int item : buckets.get(i)) {
                    if (index == k)
                        break;

                    res[index] = item;
                    index++;
                }
            }
        }

        return res;
    }
}